stream_consumer.go

```diff
@@ -31,10 +28,54 @@ type StreamEndpointConfig struct {
 	backoffMaxDelay time.Duration
 }
 
-type Consumer struct {
-	StreamName string
-	EvtChan    chan *stream.Event
+type StreamConsumer interface {
+	StreamName() string
+	EvtChan() chan *stream.Event
+	Stop()
+	streamEndpoint() *StreamEndpoint
+}
+
+type registeredConsumer struct {
+	StreamConsumer
+	g *Gaz
+}
+
+func (c *registeredConsumer) Stop() {
+	c.StreamConsumer.Stop()
+	c.g.stopConsumer(c)
+}
+
+type consumer struct {
+	endpoint   *StreamEndpoint
+	streamName string
+	evtChan    chan *stream.Event
 	config     *ConsumerConfig
+	mu         sync.RWMutex
+	stopped    bool
+}
+
+func (c *consumer) streamEndpoint() *StreamEndpoint {
+	return c.endpoint
+}
+
+func (c *consumer) StreamName() string {
+	return c.streamName
+}
+
+func (c *consumer) EvtChan() chan *stream.Event {
+	return c.evtChan
+}
+
+func (c *consumer) Stop() {
+	c.mu.Lock()
```

Comment:
I think you don't need a RWMutex here, a simple `sync/atomic` on a int would be enough

Reply:
yes, it will be better, specially now that we check it before each value reception on the streams